### ISSUE TITLE
GPU not working much

### ISSUE THREAD

**Issue by lorenzoFabbri** [skorch-dev/skorch#498]
I'm new to Skorch and today I tried to use it to fine tuning a pretrained ResNet18.
I noticed that the time per epoch is rather high compared to pure PyTorch. And I noticed something weird: I check the status of my machine with Glances and the `proc` field of the GPU status is very often close to 0%, while memory consumption is around 100%. Why is this happening? Is it normal?

**Comment by BenjaminBossan** [comment]
Could you provide a minimal code example to reproduce the issue? Otherwise, it's hard to tell.

**Comment by lorenzoFabbri** [comment]
Model:
`
class Net(nn.Module):
    
    def __init__(self):
        super().__init__()
        
        model = models.resnet18(pretrained=True)
        
        model.conv1 = nn.Conv2d(6, 64, 7, 2, 3)
        torch.nn.init.xavier_uniform_(model.conv1.weight)
        
        model.fc = nn.Linear(model.fc.in_features, NUM_CLASSES)
        torch.nn.init.xavier_uniform_(model.fc.weight)
        
        self.model = model
        
    def forward(self, x):
        
        return self.model(x)
`

Callbacks:
`
lr_scheduler = LRScheduler(
    policy='StepLR', step_size=10, gamma=0.1)

freezer = Freezer(lambda x: not x.startswith(('model.conv1', 'model.fc')))

train_acc = EpochScoring(scoring='accuracy', on_train=True, 
                         name='train_acc')

callbacks = [lr_scheduler, train_acc]
`

Fit:
`
net = NeuralNetClassifier(
    Net, 
    criterion=nn.CrossEntropyLoss, 
    lr=0.1, 
    batch_size=32, 
    max_epochs=50, 
    module_output_features=NUM_CLASSES, 
    optimizer=optim.SGD, 
    iterator_train_shuffle=True, 
    iterator_train_num_workers=4, 
    iterator_valid_shuffle=True, 
    iterator_valid_num_workers=4, 
    train_split=predefined_split(val_set), 
    callbacks=callbacks, 
    device='cuda')
`

I have a custom dataset generator, which is quite long to post here.
I also noticed that if I train it from scratch, it fluctuates a lot between 0% and 100%, while memory consumption remains constant.

**Comment by BenjaminBossan** [comment]
It is unfortunately quite hard to debug this issue without being able to reproduce it.

> I noticed that the time per epoch is rather high compared to pure PyTorch

How much of a difference do you see? A little bit of overhead is expected, but it should only matter for very small architectures, so that's unlikely here.

Often when I encounter similar discrepancies, it is because the skorch code is actually doing more work than the pure pytorch code, e.g. callbacks and performing the internal validation. If that is the case for you, try to turn off as much of these extras as possible, set everything else the same (e.g. batch size), and then compare the times.

> 
    iterator_train_shuffle=True, 
    iterator_train_num_workers=4, 
    iterator_valid_shuffle=True, 
    iterator_valid_num_workers=4, 

This code is faulty, there should be a double-underscore between the main argument and the sub-arguments, e.g. `iterator_train__num_workers=4`. Please try if changing this already helps.

**Comment by lorenzoFabbri** [comment]
I’ll fix those lines then and try again.
I would say it’s twice as slow, at the moment.

**Comment by thomasjpfan** [comment]
When there are not many weights to update, then skorch will have more of the overhead. Out of curiosity, what is the time spent per epoch with skorch and in pure PyTorch?

**Comment by BenjaminBossan** [comment]
@lorenzoFabbri Did you make any progress?

**Comment by lorenzoFabbri** [comment]
Sorry, I tried to experiment a little.
I fixed my faulty code and tried to train my model from scratch: it went from 1400 to 700 seconds per epoch. I also tried pure PyTorch and it seems comparable. It takes to much to have proper statistics, though. And I did not have time to test the pretrained model.
I still notice that I takes a lot of time to transfer data: right now my custom dataset uses a CSV file to find the paths to the single channel images to concatenate into 6-channel images. I believe this is not optimal. So probably that’s why the GPU is not working *constantly*.
I think we can close this issue. Thanks for the support.
